addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js

```diff
@@ -1476,10 +1477,13 @@ export class OdooEditor extends EventTarget {
         });
         if (!range) return;
         let { startContainer: start, startOffset, endContainer: end, endOffset } = range;
+        const startBlock = closestBlock(start);
+        const endBlock = closestBlock(end);
         const doJoin =
-            (closestBlock(start) !== closestBlock(range.commonAncestorContainer) ||
-            closestBlock(end) !== closestBlock(range.commonAncestorContainer))
-            && (closestBlock(start).tagName !== 'TD' && closestBlock(end).tagName !== 'TD');
+            (startBlock !== closestBlock(range.commonAncestorContainer) ||
+                endBlock !== closestBlock(range.commonAncestorContainer))
+            && (startBlock.tagName !== 'TD' && endBlock.tagName !== 'TD')
+            && (firstLeaf(startBlock) !== start || lastLeaf(endBlock) !== end);
```

Comment:
Style suggestion for the sake of clarity (`!(a && b)` tends to be more intuitive than `!a || !b`):

```suggestion
            && !(firstLeaf(startBlock) === start && lastLeaf(endBlock) === end);
```
I would also add a comment above it as @shsa-odoo suggested.